### PR TITLE
[SPARK-24849][SPARK-24911][SQL] Converting a value of StructType to a DDL string

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to extend the `StructType`/`StructField` classes by new method `toDDL` which converts a value of the `StructType`/`StructField` type to a string formatted in DDL style. The resulted string can be used in a table creation.

The `toDDL` method of `StructField` is reused in `SHOW CREATE TABLE`. In this way the PR fixes the bug of unquoted names of nested fields.

## How was this patch tested?

I add a test for checking the new method and 2 round trip tests: `fromDDL` -> `toDDL` and `toDDL` -> `fromDDL`

### CODE REVIEW

**Comment by maropu** [comment]
How about the case where a column name has special characters that should be backquoted, e.g., 'aaa:bbb'?

**Comment by MaxGekk** [comment]
@maropu I added quoting of column names

**Comment by maropu** [comment]
(As I described in the jira) What's this func is used for? Is this related to the other work?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 1]
```diff
@@ -436,6 +436,14 @@ object StructType extends AbstractDataType {
    */
   def fromDDL(ddl: String): StructType = CatalystSqlParser.parseTableSchema(ddl)
 
+  /**
+   * Converts a value of StructType to a string in DDL format. For example:
+   * `StructType(Seq(StructField("a", IntegerType)))` should be converted to `a int`
+   */
+  def toDDL(struct: StructType): String = {
+    struct.map(field => s"${quoteIdentifier(field.name)} ${field.dataType.sql}").mkString(",")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala]: Can this also handle the special character ('\n', '\t', '\', ...) that needs an escape?

**Comment by MaxGekk** [comment]
@hvanhovell Could you look at the PR please.

**Comment by MaxGekk** [comment]
> (As I described in the jira) What's this func is used for?

@maropu I answered in JIRA, please, look at it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala]: We use similar code in `SHOW CREATE TABLE`: https://github.com/apache/spark/blob/bac50aa37168a7612702a4503750a78ed5d59c78/sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala#L999

**Review by HyukjinKwon — COMMENTED** [review]
@MaxGekk, is the purpose of this API is to have `a int` instead of `struct<a: int>`?

You can use `struct.map(field => s"${quoteIdentifier(field.name)} ${field.dataType.sql}").mkString(",")` in the application code or you could simply:

```scala
scala> DataType.fromDDL(StructType.fromDDL("a int").catalogString)
res4: org.apache.spark.sql.types.DataType = StructType(StructField(a,IntegerType,true))
```

**Comment by HyukjinKwon** [comment]
We should reduce APIs exposed .. there are already a hell of a lot..

**Comment by MaxGekk** [comment]
> is the purpose of this API is to have a int instead of struct<a: int>

Basically, yes. All those methods `simpleString()`, `catalogString()`, `sql()` return `struct< ... : ...>` which is impossible to uses in a table creation because of `:` and `struct`

> We should reduce APIs exposed .. there are already a hell of a lot..

I believe we should reduce user's/customer's pain first of all. I think the function which is opposite to `fromDLL` looks pretty natural here.

**Comment by HyukjinKwon** [comment]
Does the alternative code I posted above work in that case? If so, sounds we are adding an API just for consistency

**Comment by MaxGekk** [comment]
> Does the alternative code I posted above work in that case?
> You can use struct.map(field => s"${quoteIdentifier(field.name)} ${field.dataType.sql}").mkString(",") in the application code

It is actually the same as in the PR. For sure, it works ( I added tests for that ;-) )

> or you could simply:
> scala> DataType.fromDDL(StructType.fromDDL("a int").catalogString)

Sorry, I didn't catch the example. Just in case, I need to read one avro file from a folder, take a schema from it in DDL format and create table by using the string. This is the use case. 

> If so, sounds we are adding an API just for consistency ...

Not only, we are adding the method to cover the use case which I described above. And for consistency too. Having useful API, isn't it beautiful?

**Comment by HyukjinKwon** [comment]
Ah, I misunderstood then. Thing is, `fromDDL` was added because it's needed other APIs rather then it's own purpose. I wonder how commonly it will be used to be honest. I haven't seen such requests before frankly.

In my case, I usually leave mailing list or stackoverflow questions as references usually though. It's one liner code and I think we haven't added APIs when the workarounds are easy usually so far.

**Comment by MaxGekk** [comment]
> Thing is, fromDDL was added because it's needed other APIs rather then it's own purpose. 

The `toDDL` function can be reused from `SHOW CREATE TABLE` too: https://github.com/apache/spark/blob/bac50aa37168a7612702a4503750a78ed5d59c78/sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala#L999

> I wonder how commonly it will be used to be honest. I haven't seen such requests before frankly.

Frankly speaking I cannot say there is high demand for the method from our users. 

> It's one liner code and I think we haven't added APIs when the workarounds are easy usually so far. 

Do you think I should close the PR?

**Comment by HyukjinKwon** [comment]
To me yup, but if you are in doubt, I am perfectly okay with waiting some more days and see if other opinions arrive.

**Comment by rxin** [comment]
Should we do schema.toDDL, or StructType.toDDL(schema)?

**Comment by gatorsmile** [comment]
schema.toDDL is more friendly.

**Comment by MaxGekk** [comment]
> Should we do schema.toDDL, or StructType.toDDL(schema)?
> schema.toDDL is more friendly.

For sure, `schema.toDDL` looks more natural. I put `toDDL` to the `StructType object` only to have it closer to `fromDDL` and to do not pollute the namespace of the `StructType class` which already has inherited from `DataType` similar methods `simpleString()`, `catalogString()` and `sql()`

**Comment by MaxGekk** [comment]
@gatorsmile @rxin I moved `toDDL` to the `StructType` class and reused it from `SHOW CREATE TABLE` (fixed the bug of unquoted field names). Please, have take a look at the PR.

**Comment by gatorsmile** [comment]
@MaxGekk Thanks for fixing the issue in `SHOW CREATE TABLE`! Could you open a JIRA and add the JIRA number to this PR. Also, could you add the failed case in ShowCreateTableSuite?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 2]
```diff
@@ -155,6 +155,18 @@ package object util {
 
   def toPrettySQL(e: Expression): String = usePrettyExpression(e).sql
 
+
+  def escapeSingleQuotedString(str: String): String = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: Why this is moved under `catalyst`?

**Comment by HyukjinKwon** [comment]
I am okay now just for clarification ~

**Comment by gatorsmile** [comment]
It is nice to have. Actually, I believe we need to fix the bug in `SHOW CREATE TABLE`, which is widely used.

**Comment by HyukjinKwon** [comment]
yup the current change sounds okay.

**Comment by gatorsmile** [comment]
@MaxGekk Please include the test case for SHOW CREATE TABLE. Thanks!

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master

**Pull request closed without merging**